Allow non-string variables

akloeckner opened the pull request:
This change makes non-string variables available to a pilight binary_sensor.

Background: The payload_on parameter is a string. If a pilight sensor sends a numeric variable, this variable not usable for the binary sensor, because a numeric veriable never is equal to a string as per the == operator.

Change: The pilight variable is converted to a string before comparison. This preserves backward compatibility of configurations.

## Description:

**Related issue (if applicable):** https://github.com/home-assistant/home-assistant/issues/17882

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** none

## Example entry for `configuration.yaml` (if applicable):
```yaml
binary_sensor:
  - platform: pilight
    variable: 'battery'
    payload_on:  '1.0'
```
or
```yaml
binary_sensor:
  - platform: pilight
    variable: 'battery'
    payload_on:  1.0
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

akloeckner committed:
Allow non-string variables 

This change makes non-string variables available to a pilight binary_sensor.

Background: The payload_on parameter is a string. If a pilight sensor sends a numeric variable, this variable not usable for the binary sensor, because a numeric veriable never is equal to a string as per the == operator.

Change: The pilight variable is converted to a string before comparison. This preserves backward compatibility of configurations.

akloeckner commented:
In principle you're right. But I don't know what the valid types are.  :-) And I have no idea how to allow multiple types.

As a side note, I also wouldn't trust testing floating point numbers (if they are a valid type) for equality.  For this, a new threshold comparison option might be the more appropriate solution.

Maybe there is a way to keep these thoughts as notes in the comments? What is your policy about that?

fabaff commented:
Details about the validation: https://developers.home-assistant.io/docs/en/development_validation.html. Only useful if you know what the types are :wink: 

There is no policy about comments. If it's valuable information at it or add a link to the source. Otherwise one needs to browse through the Pull requests.

balloob reviewed (changes requested):
Adjust the config validation to not convert user submitted values to string instead.

akloeckner committed:
Annotate changed lines

More proper solutions can be found in the PR: https://github.com/home-assistant/home-assistant/pull/17870

akloeckner commented:
Sorry, I don't get, how this would be done. I just annotated the code to make it clear for someone who can do this in the next step. For now, it would be a working fix for me without any downsides I can see.

balloob commented:
You need to change the order. vol.Any will return the first one that matches. `cv.string` will always be able to convert the value to a string. Same with float and int, int before float.
